horovod/run/gloo_run.py

```diff
@@ -234,79 +184,132 @@ def _exec_command(command, alloc_info, event):
-    # kill their corresponding middleman processes so the jobs can be killed as well.
-    alloc_info_to_command = _alloc_info_to_command_fn(run_command, env)
-    args_list = [[alloc_info_to_command(alloc_info), alloc_info, event]
-                 for alloc_info in host_alloc_plan]
+def launch_gloo(command, exec_command, settings, nics, env, server_ip):
+    """
+    Launches the given command multiple times using gloo.
+    Each command is launched via exec_command.
 
+    :param command: command to launch
+    :param exec_command: means to execute a single command
+    :param settings: settings for the distribution
+    :param nics: common interfaces
+    :param env: environment to use
+    :param server_ip: ip to use for rendezvous server
+    """
     # Make the output directory if it does not exist
     if settings.output_filename:
         _mkdir_p(settings.output_filename)
 
+    # start global rendezvous server and get port that it is listening on
+    rendezvous = RendezvousServer(settings.verbose)
+
+    # allocate processes into slots
+    hosts = parse_hosts(settings.hosts)
+    host_alloc_plan = get_host_assignments(hosts, settings.num_proc)
+
+    # start global rendezvous server and get port that it is listening on
+    global_rendezv_port = rendezvous.start_server()
+    rendezvous.httpd.init(host_alloc_plan)
+    run_command = get_run_command(command, server_ip, nics, global_rendezv_port)
+
+    slot_info_to_command = _slot_info_to_command_fn(run_command, env)
+    event = register_shutdown_event()
+    args_list = [[slot_info_to_command(slot_info), slot_info, [event]]
+                 for slot_info in host_alloc_plan]
+
     # If an error occurs in one thread, entire process will be terminated.
     # Otherwise, threads will keep running.
-    res = threads.execute_function_multithreaded(exec_command, args_list, block_until_all_done=True)
+    res = threads.execute_function_multithreaded(exec_command,
+                                                 args_list,
+                                                 block_until_all_done=True)
 
     for name, value in sorted(res.items(), key=lambda item: item[1][1]):
         exit_code, timestamp = value
         if exit_code != 0:
-            raise RuntimeError('Gloo job detected that one or more processes exited with non-zero '
+            raise RuntimeError('Horovod detected that one or more processes exited with non-zero '
                                'status, thus causing the job to be terminated. The first process '
                                'to do so was:\nProcess name: {name}\nExit code: {code}\n'
                                .format(name=name, code=exit_code))
 
 
-def gloo_run(settings, remote_host_names, nics, env, server_ip, command):
+def _get_min_start_hosts(settings):
+    # This function exists for the purpose of mocking in tests
+    return 2 if settings.max_np > settings.num_proc and not settings.nics else 1
```

Comment:
`settings.max_np` can be unbound (see https://github.com/horovod/horovod/pull/1849/files#diff-0bc78cae4a6c24827df277600992eedeR358-R360) and `None > int` throws `TypeError: '>' not supported between instances of 'NoneType' and 'int'`.

Should be `if is_auto_scale and not settings.nics` with `is_auto_scale = settings.min_np < settings.num_proc or settings.max_np is None or settings.max_np > settings.num_proc` (assuming `settings.min_np` cannot be `None` here).

Reply:
That's fair.  I've also changed behavior to raise an error if the user attempts to run in fault tolerance mode with only one host.